### PR TITLE
CBG-876: Include Bearer prefix when sending token base authentication…

### PR DESCRIPTION
**Pull request by sarathkumarsivan**
- Move getRSASigner() from util_testing.go to oidc_test.go.
- Include Bearer prefix when sending token base authentication request in OIDC E2E tests.
- Attach public keys and signer to the mockAuthServer.
- Add keysHandler to expose the public keys to validate a JSON Web Token (JWT) issued.
- Include makeToken method on mockAuthServer to creates a default token.
- E2E test that checks OpenID Connect Implicit Flow.

### CODE REVIEW

File: rest/oidc_api_test.go [hunk 1]
```diff
@@ -758,3 +806,113 @@ func refreshProviderConfig(providers auth.OIDCProviderMap, issuer string) {
...
+			}
+			require.Equal(t, http.StatusOK, response.StatusCode)
+			assert.Equal(t, "application/json", response.Header.Get("Content-Type"))
+
+			var responseBody map[string]interface{}
+			require.NoError(t, err, json.NewDecoder(response.Body).Decode(&responseBody))
+			sessionCookie := getCookie(response.Cookies(), auth.DefaultCookieName)
+			require.NotNil(t, sessionCookie, "No session cookie found")
+			sessionExpiry := int(sessionCookie.Expires.Sub(time.Now()).Seconds())
+			assert.True(t, 86395 <= sessionExpiry && 86400 >= sessionExpiry, "session expiry is not within 5 seconds of 24 hours")
```
Comment (adamcfraser) [inline rest/oidc_api_test.go]: 24h is the default SG session timeout (kDefaultSessionTTL).   Is that what we want to be testing here?  It's a bit unexpected in  what appears to be an OIDC test.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: rest/oidc_api_test.go [hunk 2]
```diff
@@ -259,6 +281,30 @@ func (s *mockAuthServer) tokenHandler(w http.ResponseWriter, r *http.Request) {
 	renderJSON(w, r, http.StatusOK, response)
 }
 
+// makeToken creates a default token with an expiry of 5 minutes.
+func (s *mockAuthServer) makeToken() (string, error) {
```
[the remark this answers is not in the document]
Comment (sarathkumarsivan) [reply rest/oidc_api_test.go]: Makes sense. The makeToken method has been refactored to accept claimSet which contains both primary and secondary claims. Now it is convenient to adjust claims from the caller for specific use cases. Also included additional unit test cases to cover the edge cases.

**Review by sarathkumarsivan — COMMENTED** [review]
- Added comments against BearerToken stating that it is used for setting JWT token type as well as the prefix for the Authorization HTTP header.
- Refactored makeToken method to keep one property per line while building the claims for making auth token.
- Refactored getCookie function to use enhanced for loop.
- Since we're already checking whether the DefaultCookieName exists, skip the check for kDefaultSessionTTL and make sure the response body has the expected details like username, channels, authentication_handlers etc.
- Refactored makeToken method to accept claimSet which contains both primary and secondary claims to make it convenient to adjust claims from the caller for specific use cases. 
- Included additional unit test cases to cover the edge cases.

File: rest/oidc_api_test.go [hunk 3]
```diff
@@ -692,7 +740,7 @@ func TestOpenIDConnectAuth(t *testing.T) {
 			dbEndpoint := mockSyncGatewayURL + "/" + restTester.DatabaseConfig.Name
 			request, err = http.NewRequest(http.MethodGet, dbEndpoint, nil)
 			require.NoError(t, err, "Error creating new request")
-			request.Header.Add("Authorization", authResponseActual.IDToken)
+			request.Header.Add("Authorization", BearerToken+" "+authResponseActual.IDToken)
```
[the remark this answers is not in the document]
Comment (adamcfraser) [reply rest/oidc_api_test.go:760]: I was wondering if that's what was going on.  Filed CBG-879 to force tests to opt-in to GUEST usage.

**Review by adamcfraser — APPROVED** [review]
